### ISSUE TITLE
Strengthen type for partition

### ISSUE THREAD

**Issue by bennn** [racket/typed-racket#138]
The current types for partition are:
```
(-> a Any) (Listof a) -> (values (Listof a) (Listof a))
(-> Any Boolean : a) (Listof b) -> (values (Listof a) (Listof b))
```
This second type is too weak. The type of the second list returned should express that the filter failed on all its elements. For example, I'd like this example to typecheck:

```
#lang typed/racket/base

(require (only-in racket/list partition))

(: xs (Listof (U Boolean Char)))
(define xs '())

(let-values: ([([bools : (Listof Boolean)]
                [chars : (Listof Char)])
               (partition boolean? xs)])
  (void))
```

---

I was hoping to make this change myself, but I don't know how to express the filter. The closest I've come up with is to change the type of partition to use a `not-filter`.
```
[partition                                                                      
  (-poly (a b) (cl->* (-> (-> a Univ) (-lst a) (-values (list (-lst a) (-lst a))))
               (-> (make-pred-ty a) (-lst b) (-values (list (-lst a) (-lst (-not-filter a (make-Path null (list 0 0)))))))))]
```
But I doubt this compiles, and the `not-filter` should mention `b`.

Can someone with more experience help me?

**Comment by samth** [comment]
I don't think we can do this currently, but it will be possible when @andmkent lands #45.

**Comment by AlexKnauth** [comment]
For the type of partition, would it make sense to change it to this:
```
(All (a b)
 (case->
  (-> (-> a Any) (Listof a) (values (Listof a) (Listof a)))
  (-> (-> b Any : #:+ a) (Listof b) (values (Listof a) (Listof b)))))
```

**Comment by bennn** [comment]
@AlexKnauth 
The `#:+ a` is good, but I also want a `#:- c`. Calling `partition boolean?` on a list of `U Boolean Char` should return one `Listof Boolean` and one `Listof Char`.

**Comment by AlexKnauth** [comment]
The problem with that is that currently this doesn't work:
```racket
#lang typed/racket
(: f : [(U Boolean Char) -> Boolean : #:+ Boolean #:- Char])
(define f boolean?)
```
Actually not even this works:
```racket
#lang typed/racket
(: f : [(U Boolean Char) -> Boolean : #:+ Boolean #:- Char])
(define (f x)
  (and (boolean? x) (not (char? x))))
```
So even if the type of partition supported that, it wouldn't help much because `boolean?` doesn't have a type that's a subtype of `[(U Boolean Char) -> Boolean : #:+ Boolean #:- Char]`, and it would be hard to make your own predicate with that type, and not very convenient anyway.  What you could do though, is define your own partition-like function that takes two predicates, like this: https://github.com/AlexKnauth/tic-tac-toe/blob/master/tic-tac-toe/utils/partition2.rkt
(and yes it's typed racket code)
